mono/tests/threadpool-exceptions2.cs

```diff
@@ -2,38 +2,46 @@
 using System.Threading;
 
 class Test {
-	static object monitor;
-	static int return_value = 2;
+
+	static ManualResetEvent mre = new ManualResetEvent (false);
 
 	static int Main ()
 	{
-		monitor = new object ();
 		AppDomain.CurrentDomain.UnhandledException += OnUnhandledException;
 		WaitCallback wcb = new WaitCallback ((a) => {
 			throw new Exception ("From the threadpoool");
 		});
 		wcb.BeginInvoke (wcb, OnCBFinished, null);
-		lock (monitor) {
-			Monitor.Wait (monitor);
-		}
-		Thread.Sleep (1000);
-		return 1;
+
+		if (!mre.WaitOne (10000))
+			return 2;
+
+		GC.Collect ();
+		GC.WaitForPendingFinalizers ();
+
+		/* expected exit code: 255 */
+		Thread.Sleep (10000);
+		return 0;
 	}
 
 	static void OnUnhandledException (object sender, UnhandledExceptionEventArgs e)
 	{
 		string str = e.ExceptionObject.ToString ();
-		if (str.IndexOf ("From the threadpool") != -1)
-			return_value = 3;
-		lock (monitor) {
-			Monitor.Pulse (monitor);
+		if (!str.Contains ("From OnCBFinished")) {
+			Environment.Exit (3);
+			return;
+		}
+
+		if (!e.IsTerminating) {
+			Environment.Exit (4);
+			return;
 		}
-		Environment.Exit (return_value);
+
+		mre.Set ();
```

Comment:
Why not having the same `Environment.Exit (0)` as the other tests?

Reply:
Because this is exit path for unhandled exception